plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Costing.hs

```diff
@@ -0,0 +1,425 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeApplications  #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Evaluation.Builtins.Costing where
+
+import PlutusCore.Evaluation.Machine.CostStream
+import PlutusCore.Evaluation.Machine.ExBudget
+import PlutusCore.Evaluation.Machine.ExBudgetStream
+import PlutusCore.Evaluation.Machine.ExMemory
+import PlutusCore.Evaluation.Machine.ExMemoryUsage
+
+import Data.Bifunctor
+import Data.Coerce
+import Data.Int
+import Data.List
+import Data.Maybe
+import Data.SatInt
+import Test.QuickCheck.Gen
+import Test.Tasty
+import Test.Tasty.QuickCheck
+
+deriving newtype instance Foldable NonEmptyList  -- QuickCheck...
+
+eqCostStream :: CostStream -> CostStream -> Bool
+eqCostStream (CostLast cost1) (CostLast cost2) = cost1 == cost2
+eqCostStream (CostCons cost1 costs1) (CostCons cost2 costs2) =
+    cost1 == cost2 && eqCostStream costs1 costs2
+eqCostStream _ _ = False
+
+eqExBudgetStream :: ExBudgetStream -> ExBudgetStream -> Bool
+eqExBudgetStream (ExBudgetLast budget1) (ExBudgetLast budget2) = budget1 == budget2
+eqExBudgetStream (ExBudgetCons budget1 budgets1) (ExBudgetCons budget2 budgets2) =
+    budget1 == budget2 && eqExBudgetStream budgets1 budgets2
+eqExBudgetStream _ _ = False
+
+fromCostList :: NonEmptyList CostingInteger -> CostStream
+fromCostList (NonEmpty [])               = error "cannot be empty"
+fromCostList (NonEmpty (cost0 : costs0)) = go cost0 costs0 where
+    go cost []              = CostLast cost
+    go cost (cost' : costs) = CostCons cost $ go cost' costs
+
+toCostList :: CostStream -> NonEmptyList CostingInteger
+toCostList = NonEmpty . go where
+    go (CostLast cost)       = [cost]
+    go (CostCons cost costs) = cost : go costs
+
+toExBudgetList :: ExBudgetStream -> NonEmptyList ExBudget
+toExBudgetList = NonEmpty . go where
+    go (ExBudgetLast budget)         = [budget]
+    go (ExBudgetCons budget budgets) = budget : go budgets
+
+-- >>> magnitudes
+-- [(0,10),(11,100),(101,1000),(1001,10000),(10001,100000),(100001,1000000),(1000001,10000000),(10000001,100000000),(100000001,1000000000),(1000000001,10000000000),(10000000001,100000000000),(100000000001,1000000000000),(1000000000001,10000000000000),(10000000000001,100000000000000),(100000000000001,1000000000000000),(1000000000000001,10000000000000000),(10000000000000001,100000000000000000),(100000000000000001,1000000000000000000),(1000000000000000001,9223372036854775807)]
```

Comment:
> If you want to generate numbers between 1 and b whose lengths are uniformly distributed then you could generate uniform numbers in the range `[0,log b)` (say log_10, but it doesn't matter) and then exponentiate those, so something like

This would only generate nice round numbers, right? I don't want to discriminate against the ugly ones!

> You could also bias the numbers towards the endpoints by taking a normal distribution, chopping it in half, and swapping the halves around (there are 10,000 samples here):

That's clever. For now I simply skew generation of numbers towards smaller ones, so that when we generate streams of those we get a pretty nice distribution like this:

```
        +++ OK, passed 10000 tests:
         7.54% 1000000000001 - 10000000000000
         7.31% 100000000001 - 1000000000000
         7.06% 10000000000001 - 100000000000000
         6.99% 100000000000001 - 1000000000000000
         6.45% 10000000001 - 100000000000
         6.16% 1000000001 - 10000000000
         6.13% 1000000000000001 - 10000000000000000
         5.35% 10000000000000001 - 100000000000000000
         5.22% 100000001 - 1000000000
         4.77% 1000000000000000001 - 9223372036854775807
         4.77% 10000001 - 100000000
         4.76% 0 - 10
         4.35% 1000001 - 10000000
         4.21% 11 - 100
         4.13% 100001 - 1000000
         3.96% 100000000000000001 - 1000000000000000000
         3.75% 101 - 1000
         3.57% 10001 - 100000
         3.52% 1001 - 10000
```

I do think we can improve it, but it's perhaps not a priority right now.

Reply:
> This would only generate nice round numbers, right? I don't want to discriminate against the ugly ones!

No, I don't think so.  Generate floats/doubles in the range [0, log maxBound) and then call `exp` and take the floor.  I think that'll give you numbers like you're currently getting but more smoothly distributed.  I'll try that in your branch and see how the numbers compare with the ones above.

Reply:
Hmmm.  Maybe not.  I tried
```
    arbitrary = (unsafeToSatInt . floor . exp) <$> (choose (0::Double, log $ fromSatInt maxBound))
```
and got 
```
     distribution of the generated CostStream values:                         OK (0.20s)
        +++ OK, passed 10000 tests:
        53.09% 1000000000000000001 - 9223372036854775807
        12.47% 100000000000000001 - 1000000000000000000
         6.19% 10000000000000001 - 100000000000000000
         4.70% 1000000000000001 - 10000000000000000
         3.34% 100000000000001 - 1000000000000000
         2.81% 10000000000001 - 100000000000000
         2.24% 1000000000001 - 10000000000000
         1.73% 10000000001 - 100000000000
         1.57% 1000000001 - 10000000000
         1.50% 100000000001 - 1000000000000
         1.38% 1000001 - 10000000
         1.35% 100000001 - 1000000000
         1.33% 10000001 - 100000000
         1.19% 10001 - 100000
         1.12% 1001 - 10000
         1.11% 0 - 10
         1.01% 100001 - 1000000
         0.95% 101 - 1000
         0.92% 11 - 100
```
It could be the case that `choose` isn't producing uniformly distributed numbers though.  I'll have to go and remind myself how QuickCheck generators work.

Reply:
>  Generate floats/doubles

Ah right, I missed that those were doubles rather than integers.

> It could be the case that choose isn't producing uniformly distributed numbers though. I'll have to go and remind myself how QuickCheck generators work.

No, I think it's the expected result. Note that the current generator of `SatInt`s is skewed towards generating small values, so that the generator for `CostStream` is less skewed towards bigger sums:

```
-- | Generate asymptotically bigger 'SatInt's with exponentially lower chance. This is in order to
-- make the generator of 'CostStream' produce streams whose sums are more or less evenly distributed
-- across 'magnitudes'.
instance Arbitrary SatInt where
    arbitrary = frequency . zip freqs . reverse $ map chooseSatInt magnitudes where
        freqs = map floor $ iterate (* 1.3) (2 :: Double)
```